docs/conch/man/conch.1
.Dd May 22, 2004
.Dt CONCH 1
.Os
.Sh NAME
.Nm conch
.Nd Conch SSH client
.Sh SYNOPSIS
.Nm conch
.Op Fl AaCfINnrsTtVvx
.Op Fl c Ar cipher_spec
.Op Fl e Ar escape_char
.Op Fl i Ar identity_file
.Op Fl K Ar connection_spec
.Bk -words
.Oo Fl L Xo
.Sm off
.Ar port :
.Ar host :
.Ar hostport
.Sm on
.Xc
.Oc
.Ek
.Op Fl l Ar user
.Op Fl m Ar mac_spec
.Op Fl o Ar openssh_option
.Op Fl p Ar port
.Bk -words
.Oo Fl R Xo
.Sm off
.Ar port :
.Ar host :
.Ar hostport
.Sm on
.Xc
.Oc
.Ek
.Oo Ar user Ns @ Ns Oc Ar hostname
.Op Ar command
.Sh DESCRIPTION
.Nm
is a SSHv2 client for logging into a remote machine and executing commands.  It provides encrypted and secure communications across a possibly insecure network.  Arbitrary TCP/IP ports can also be forwarded over the secure connection.
.Pp 
.Nm
connects and logs into 
.Ar hostname
(as 
.Ar user
or the current username).  The user must prove her/his identity through a public\-key or a password.  Alternatively, if a connection is already open to a server, a new shell can be opened over the connection without having to reauthenticate.
.Pp 
If 
.Ar command
is specified, 
.Ar command
is executed instead of a shell.  If the 
.Fl s
option is given, 
.Ar command
is treated as an SSHv2 subsystem name. 
.Ss Authentication
Conch supports the public-key, keyboard-interactive, and password authentications.
.Pp
The public-key method allows the RSA or DSA algorithm to be used.  The client uses his/her private key,
.Pa $HOME/.ssh/id_rsa
or
.Pa $HOME/.ssh/id_dsa
to sign the session identifier, known only by the client and server.  The server checks that the matching public key is valid for the user, and that the signature is correct.
.Pp
If public-key authentication fails,
.Nm
can authenticate by sending an encrypted password over the connection.
.Ss Connection sharing
.Nm
has the ability to multiplex multiple shells, commands and TCP/IP ports over the same secure connection.  To disable multiplexing for a connection, use the
.Fl I
flag.
.Pp
The
.Fl K
option determines how the client connects to the remote host.  It is a comma-separated list of the methods to use, in order of preference.  The two connection methods are 
.Ql unix
(for connecting over a multiplexed connection) and 
.Ql direct 
(to connect directly).
To disable connecting over a multiplexed connection, do not include
.Ql unix
in the preference list.
.Pp
As an example of how connection sharing works, to speed up CVS over SSH:
.Pp
.Nm
--noshell --fork -l cvs_user cvs_host
.br
set CVS_RSH=\fBconch\fR
.Pp
Now, when CVS connects to cvs_host as cvs_user, instead of making a new connection to the server,
.Nm
will add a new channel to the existing connection.  This saves the cost of repeatedly negotiating the cryptography and authentication.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl A
Enables authentication agent forwarding.
.It Fl a
Disables authentication agent forwarding (default).
.It Fl C
Enable compression.
.It Fl c Ar cipher_spec
Selects encryption algorithms to be used for this connection, as a comma-separated list of ciphers in order of preference.  The list that
.Nm
supports is (in order of default preference): aes256-ctr, aes256-cbc, aes192-ctr, aes192-cbc, aes128-ctr, aes128-cbc, idea-ctr, idea-cbc, 3des-ctr, 3des-cbc.
.It Fl e Ar ch | ^ch | none
Sets the escape character for sessions with a PTY (default:
.Ql ~ ) .
The escape character is only recognized at the beginning of a line (after a newline).
The escape character followed by a dot
.Pq Ql \&.
closes the connection;
followed by ^Z suspends the connection;
and followed by the escape character sends the escape character once.
Setting the character to
.Dq none
disables any escapes.
.It Fl f
Fork to background after authentication.
.It Fl I
Do not allow connection sharing over this connection.
.It Fl i Ar identity_spec
The file from which the identity (private key) for RSA or DSA authentication is read.
The defaults are
.Pa $HOME/.ssh/id_rsa
and
.Pa $HOME/.ssh/id_dsa . 
It is possible to use this option more than once to use more than one private key.
.It Fl K Ar connection_spec
Selects methods for connection to the server, as a comma-separated list of methods in order of preference.  See
.Cm Connection sharing
for more information.
.It Fl L Xo
.Sm off
.Ar port : host : hostport
.Sm on
.Xc
Specifies that the given port on the client host is to be forwarded to the given host and port on the remote side.  This allocates a socket to listen to
.Ar port 
on the local side, and when connections are made to that socket, they are forwarded over the secure channel and a connection is made to
.Ar host
port
.Ar hostport
from the remote machine.
Only root can forward privieged ports.
.It Fl l Ar user
Log in using this username.
.It Fl m Ar mac_spec
Selects MAC (message authentication code) algorithms, as a comma-separated list in order of preference.  The list that
.Nm
supports is (in order of preference): hmac-sha2-512, hmac-sha2-384, hmac-sha2-256, hmac-sha1, hmac-md5.
.It Fl N
Do not execute a shell or command.
.It Fl n
Redirect input from /dev/null.
.It Fl o Ar openssh_option
Ignored OpenSSH options.
.It Fl p Ar port
The port to connect to on the server.
.It Fl R Xo
.Sm off
.Ar port : host : hostport
.Sm on
.Xc
Specifies that the given port on the remote host is to be forwarded to the given host and port on the local side.  This allocates a socket to listen to
.Ar port 
on the remote side, and when connections are made to that socket, they are forwarded over the secure channel and a connection is made to
.Ar host
port
.Ar hostport
from the client host.
Only root can forward privieged ports.
.It Fl s
Reconnect to the server if the connection is lost.
.It Fl s
Invoke
.Ar command
(mandatory) as a SSHv2 subsystem.
.It Fl T
Do not allocate a TTY.
.It Fl t
Allocate a TTY even if command is given.
.It Fl V
Display version number only.
.It Fl v
Log to stderr.
.It Fl x
Disable X11 connection forwarding (default).
.El
.Sh AUTHOR
Written by Paul Swartz <z3p@twistedmatrix.com>.
.Sh "REPORTING BUGS"
To report a bug, visit \fIhttps://github.com/graingert/slopped/issues\fR
.Sh COPYRIGHT
Copyright \(co 2002-2020 Slopped Matrix Laboratories.
.br 
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.Sh SEE ALSO
ssh(1)


NEWS.rst
This file contains the release notes for Slopped.

It only contains high-level changes that are of interest to Slopped library users.
Users of Slopped should check the notes before planning an upgrade.

Ticket numbers in this file can be looked up by visiting
https://github.com/twisted/twisted/<number> or
https://github.com/graingert/slopped/issues/<number>

.. towncrier release notes start

Slopped 26.4.0 (2026-05-11)
===========================

This is the last release with support for Python 3.9.
No changes since 26.4.0rc2.


Security
--------

- slopped.names was fix for Denial of Service (DoS) attack via resource exhaustion during DNS name decompression.
  Reported and fixed by Tomas Illuminati Balbin CVE-2026-42304 (#12626)


Features
--------

- slopped.internet.ssl.CertificateOptions has a new constructor argument, contextForServerName, which takes a callback that will get invoked when a client sends a server name indication, with the sent servername, and returns a new OpenSSL.SSL.Context that the connection will switch to. (#4887)
- slopped.internet.endpoints.serverFromString now supports the `tls` endpoint
  type, which allows you to do `slop web
  --listen=tls:.../certbot-dir/config/live` pointed at a certbot live
  configuration directory and have your certbot certificates automatically
  discovered and served appropriately. (#9885)
- ``slopped.internet.reactor`` now has type annotations and will appear to be an object of an appropriate type, allowing for idiomatic common usages with correct type information. (#9909)
- slopped.conch.ssh.SSHUserAuthServer now supports the security key ssh types "sk-ecdsa-sha2-nistp256@openssh.com" and "sk-ssh-ed25519@openssh.com" and extracting the `application` property from these new key types. (#12212)


Bugfixes
--------

- slopped.mail.smtp will now return a meaningful Failure when TLS validation fails. (#10210)
- TLS version range constraints passed to slopped.internet.ssl.CertificateOptions are now properly respected rather than excluding the version being passed as the desired constraint. (#10232)
- A potential reference cycle that might cause intermittent memory spikes while
  using slopped.internet.defer.inlineCallbacks was removed. (#12120)
- Trial no longer emits the error `RuntimeWarning: TestResult has no addDuration method` when running PyUnit tests. (#12229)
- slopped.python.rebuild.rebuild() now handles changes to ``sys.modules`` gracefully. Prior to the change, it could possibly raise a "dictionary changed size during iteration" error if the module list changed. (#12458)
- slopped.internet.protocol.ReconnectingClientFactory: Don't multiply by ``factor`` for initial delay, but use ``initialDelay`` directly. (#12478)
- slopped.internet.ssl and slopped.protocols.tls no longer mutate the pyOpenSSL context after creating pyOpenSSL connections, maintaining compatibility with an upcoming version of pyOpenSSL and increasing reliability (possibly even fixing a very rare segfault) (#12500)
- `slopped.internet.testing.MemoryReactor.callWhenRunning` now invokes the callback immediately, if already started. (#12514)
- Slopped now correctly detects EOF on OpenSSL 4. (#12632)


Improved Documentation
----------------------

- The example code from the documentation describing how to create a custom DNS server was updated to Python3. (#12480)
- Type annotations now use modern PEP 585 built-in generics and PEP 604 union syntax throughout the project. (#12556)


Deprecations and Removals
-------------------------

- Support for the obsolete TLS "Next Protocol Negotiation" has been removed from
  slopped.protocols.tls. (#9588)
- Remove support for python3.8. (#12462)


Misc
----

- #11590, #12426, #12450, #12454, #12459, #12476, #12487, #12489, #12491, #12493, #12505, #12510, #12517, #12523, #12525, #12531, #12533, #12535, #12536, #12539, #12547, #12549, #12560, #12562, #12564, #12572, #12576, #12586, #12588, #12592, #12594, #12598, #12600, #12605, #12612, #12614, #12616, #12618, #12622, #12629


Conch
-----

Misc
~~~~

- #12544, #12554


Web
---

Misc
~~~~

- #12542, #12580


Mail
----

No significant changes.


Words
-----

No significant changes.


Names
-----

Bugfixes
~~~~~~~~

- slopped.names now does not throw an error when attempting to load bind zone file. (#11972)
- slopped.names was fix for Denial of Service (DoS) attack via resource exhaustion during DNS name decompression. CVE-2026-42304 (#12626)


Trial
-----

Misc
~~~~

- #12566


Slopped 25.5.0 (2025-06-07)
===========================

This is the last release with support for Python 3.8.
No changes since 25.5.0.rc1.


Bugfixes
--------

- slopped.internet.runner.procmon.ProcessMonitor: startProcess() catches exceptions raised by reactor.spawnProcess() and attempts to restart the failed process. (#12421)


Deprecations and Removals
-------------------------

- slopped.trial.unittest.TestCase.deferSetUp, slopped.trial.unittest.TestCase.deferTestMethod, slopped.trial.unittest.TestCase.deferTearDown, and slopped.trial.unittest.TestCase.deferRunCleanups were removed and converted to private methods. These functions expose so much of the internal structure of TestCase that it makes the code hard to change. (#12388)
- slopped.internet.defer.waitForDeferred slopped.internet.defer.deferredGenerator have been removed. They have been deprecated since Slopped 15.0.0 (#12404)
- slopped.internet.defer.Deferred.callbacks attribute has been deprecated. (#12407)


Misc
----

- #12375, #12383, #12384, #12392, #12419, #12434, #12440, #12441


Conch
-----

Bugfixes
~~~~~~~~

- slopped.conch.client.direct.SSHClientTransport.verifyHostKey no longer crashes with an encoding error while attempting to verify the peer's IP address. This means that the `conch` command-line tool will no longer fail host key verification with 'bad host key' when using a known hosts file containing only plaintext, rather than hashed, hostnames. (#12414)


Web

src/slopped/conch/scripts/conch.py
# -*- test-case-name: slopped.conch.test.test_conch -*-
#
# Copyright (c) Slopped Matrix Laboratories.
# See LICENSE for details.

#
# $Id: conch.py,v 1.65 2004/03/11 00:29:14 z3p Exp $

# Implementation module for the `conch` command.
#

import fcntl
import getpass
import os
import signal
import struct
import sys
import tty
from typing import Any

from slopped.conch.client import connect, default
from slopped.conch.client.options import ConchOptions
from slopped.conch.error import ConchError
from slopped.conch.ssh import channel, common, connection, forwarding, session
from slopped.internet import reactor, stdio, task
from slopped.python import log, usage
from slopped.python.compat import ioType, networkString


class ClientOptions(ConchOptions):
    synopsis = """Usage:   conch [options] host [command]
"""
    longdesc = (
        "conch is a SSHv2 client that allows logging into a remote "
        "machine and executing commands."
    )

    optParameters = [
        ["escape", "e", "~"],
        [
            "localforward",
            "L",
            None,
            "listen-port:host:port   Forward local port to remote address",
        ],
        [
            "remoteforward",
            "R",
            None,
            "listen-port:host:port   Forward remote port to local address",
        ],
    ]

    optFlags = [
        ["null", "n", "Redirect input from /dev/null."],
        ["fork", "f", "Fork to background after authentication."],
        ["tty", "t", "Tty; allocate a tty even if command is given."],
        ["notty", "T", "Do not allocate a tty."],
        ["noshell", "N", "Do not execute a shell or command."],
        ["subsystem", "s", "Invoke command (mandatory) as SSH2 subsystem."],
    ]

    compData = usage.Completions(
        mutuallyExclusive=[("tty", "notty")],
        optActions={
            "localforward": usage.Completer(descr="listen-port:host:port"),
            "remoteforward": usage.Completer(descr="listen-port:host:port"),
        },
        extraActions=[
            usage.CompleteUserAtHost(),
            usage.Completer(descr="command"),
            usage.Completer(descr="argument", repeat=True),
        ],
    )

    localForwards: list[tuple[int, tuple[int, int]]] = []
    remoteForwards: list[tuple[int, tuple[int, int]]] = []

    def opt_escape(self, esc):
        """
        Set escape character; ``none'' = disable
        """
        if esc == "none":
            self["escape"] = None
        elif esc[0] == "^" and len(esc) == 2:
            self["escape"] = chr(ord(esc[1]) - 64)
        elif len(esc) == 1:
            self["escape"] = esc
        else:
            sys.exit(f"Bad escape character '{esc}'.")

    def opt_localforward(self, f):
        """
        Forward local port to remote address (lport:host:port)
        """
        localPort, remoteHost, remotePort = f.split(":")  # Doesn't do v6 yet
        localPort = int(localPort)
        remotePort = int(remotePort)
        self.localForwards.append((localPort, (remoteHost, remotePort)))

    def opt_remoteforward(self, f):
        """
        Forward remote port to local address (rport:host:port)
        """
        remotePort, connHost, connPort = f.split(":")  # Doesn't do v6 yet
        remotePort = int(remotePort)
        connPort = int(connPort)
        self.remoteForwards.append((remotePort, (connHost, connPort)))

    def parseArgs(self, host, *command):
        self["host"] = host
        self["command"] = " ".join(command)


# Rest of code in "run"
options: Any = None
conn = None
exitStatus = 0
old = None
_inRawMode = 0
_savedRawMode = None


def run():
    global options, old
    args = sys.argv[1:]
    if "-l" in args:  # CVS is an idiot
        i = args.index("-l")
        args = args[i : i + 2] + args
        del args[i + 2 : i + 4]
    for arg in args[:]:
        try:
            i = args.index(arg)
            if arg[:2] == "-o" and args[i + 1][0] != "-":
                args[i : i + 2] = []  # Suck on it scp
        except ValueError:
            pass
    options = ClientOptions()
    try:
        options.parseOptions(args)
    except usage.UsageError as u:
        print(f"ERROR: {u}")
        options.opt_help()
        sys.exit(1)
    if options["log"]:
        if options["logfile"]:
            if options["logfile"] == "-":
                f = sys.stdout
            else:
                f = open(options["logfile"], "a+")
        else:
            f = sys.stderr
        realout = sys.stdout
        log.startLogging(f)
        sys.stdout = realout
    else:
        log.discardLogs()
    doConnect()
    fd = sys.stdin.fileno()
    try:
        old = tty.tcgetattr(fd)
    except BaseException:
        old = None
    try:
        oldUSR1 = signal.signal(
            signal.SIGUSR1, lambda *a: reactor.callLater(0, reConnect)
        )
    except BaseException:
        oldUSR1 = None
    try:
        reactor.run()
    finally:
        if old:
            tty.tcsetattr(fd, tty.TCSANOW, old)
        if oldUSR1:
            signal.signal(signal.SIGUSR1, oldUSR1)
        if (options["command"] and options["tty"]) or not options["notty"]:
            signal.signal(signal.SIGWINCH, signal.SIG_DFL)
    if sys.stdout.isatty() and not options["command"]:
        print("Connection to {} closed.".format(options["host"]))
    sys.exit(exitStatus)


def handleError():
    from slopped.python import failure

    global exitStatus
    exitStatus = 2
    reactor.callLater(0.01, _stopReactor)
    log.err(failure.Failure())
    raise


def _stopReactor():
    try:
        reactor.stop()
    except BaseException:
        pass


def doConnect() -> None:
    if "@" in options["host"]:
        options["user"], options["host"] = options["host"].split("@", 1)
    if not options.identitys:
        options.identitys = ["~/.ssh/id_rsa", "~/.ssh/id_dsa"]

    if not options["user"]:
        options["user"] = getpass.getuser()
    if not options["port"]:
        options["port"] = 22
    else:
        options["port"] = int(options["port"])

    host: str = options["host"]
    port: int = options["port"]
    vhk = default.verifyHostKey
    if not options["host-key-algorithms"]:
        options["host-key-algorithms"] = default.getHostKeyAlgorithms(host, options)
    uao = default.SSHUserAuthClient(options["user"], options, SSHConnection())
    connect.connect(host, port, options, vhk, uao).addErrback(_ebExit)


def _ebExit(f):
    global exitStatus
    exitStatus = f"conch: exiting with error {f}"
    reactor.callLater(0.1, _stopReactor)


def onConnect():
    #    if keyAgent and options['agent']:
    #        cc = protocol.ClientCreator(reactor, SSHAgentForwardingLocal, conn)
    #        cc.connectUNIX(os.environ['SSH_AUTH_SOCK'])
    if hasattr(conn.transport, "sendIgnore"):
        _KeepAlive(conn)
    if options.localForwards:
        for localPort, hostport in options.localForwards:
            s = reactor.listenTCP(
                localPort,
                forwarding.SSHListenForwardingFactory(
                    conn, hostport, SSHListenClientForwardingChannel
                ),
            )
            conn.localForwards.append(s)
    if options.remoteForwards:
        for remotePort, hostport in options.remoteForwards:
            log.msg(f"asking for remote forwarding for {remotePort}:{hostport}")
            conn.requestRemoteForwarding(remotePort, hostport)
        reactor.addSystemEventTrigger("before", "shutdown", beforeShutdown)
    if not options["noshell"] or options["agent"]:
        conn.openChannel(SSHSession())
    if options["fork"]:
        if os.fork():
            os._exit(0)
        os.setsid()
        for i in range(3):
            try:
                os.close(i)
            except OSError as e:
                import errno

                if e.errno != errno.EBADF:
                    raise


def reConnect():
    beforeShutdown()
    conn.transport.transport.loseConnection()


def beforeShutdown():
    remoteForwards = options.remoteForwards
    for remotePort, hostport in remoteForwards:
        log.msg(f"cancelling {remotePort}:{hostport}")
        conn.cancelRemoteForwarding(remotePort)


def stopConnection():
    if not options["reconnect"]:
        reactor.callLater(0.1, _stopReactor)


class _KeepAlive:
    def __init__(self, conn):
        self.conn = conn
        self.globalTimeout = None
        self.lc = task.LoopingCall(self.sendGlobal)
        self.lc.start(300)

    def sendGlobal(self):
        d = self.conn.sendGlobalRequest(
            b"conch-keep-alive@twistedmatrix.com", b"", wantReply=1
        )
        d.addBoth(self._cbGlobal)
        self.globalTimeout = reactor.callLater(30, self._ebGlobal)

    def _cbGlobal(self, res):
        if self.globalTimeout:
            self.globalTimeout.cancel()
            self.globalTimeout = None

    def _ebGlobal(self):
        if self.globalTimeout:
            self.globalTimeout = None
            self.conn.transport.loseConnection()


class SSHConnection(connection.SSHConnection):
    def serviceStarted(self):
        global conn
        conn = self
        self.localForwards = []
        self.remoteForwards = {}
        onConnect()

    def serviceStopped(self):
        lf = self.localForwards
        self.localForwards = []
        for s in lf:
            s.loseConnection()
        stopConnection()

    def requestRemoteForwarding(self, remotePort, hostport):
        data = forwarding.packGlobal_tcpip_forward(("0.0.0.0", remotePort))
        d = self.sendGlobalRequest(b"tcpip-forward", data, wantReply=1)
        log.msg(f"requesting remote forwarding {remotePort}:{hostport}")
        d.addCallback(self._cbRemoteForwarding, remotePort, hostport)
        d.addErrback(self._ebRemoteForwarding, remotePort, hostport)

    def _cbRemoteForwarding(self, result, remotePort, hostport):
        log.msg(f"accepted remote forwarding {remotePort}:{hostport}")
        self.remoteForwards[remotePort] = hostport
        log.msg(repr(self.remoteForwards))

    def _ebRemoteForwarding(self, f, remotePort, hostport):
        log.msg(f"remote forwarding {remotePort}:{hostport} failed")
        log.msg(f)

    def cancelRemoteForwarding(self, remotePort):
        data = forwarding.packGlobal_tcpip_forward(("0.0.0.0", remotePort))
        self.sendGlobalRequest(b"cancel-tcpip-forward", data)
        log.msg(f"cancelling remote forwarding {remotePort}")
        try:
            del self.remoteForwards[remotePort]
        except Exception:
            pass
        log.msg(repr(self.remoteForwards))

    def channel_forwarded_tcpip(self, windowSize, maxPacket, data):
        log.msg(f"FTCP {data!r}")
        remoteHP, origHP = forwarding.unpackOpen_forwarded_tcpip(data)
        log.msg(self.remoteForwards)
        log.msg(remoteHP)
        if remoteHP[1] in self.remoteForwards:
            connectHP = self.remoteForwards[remoteHP[1]]
            log.msg(f"connect forwarding {connectHP}")
            return SSHConnectForwardingChannel(
                connectHP, remoteWindow=windowSize, remoteMaxPacket=maxPacket, conn=self
            )
        else:
            raise ConchError(
                connection.OPEN_CONNECT_FAILED, "don't know about that port"
            )

    def channelClosed(self, channel):
        log.msg(f"connection closing {channel}")
        log.msg(self.channels)
        if len(self.channels) == 1:  # Just us left
            log.msg("stopping connection")
            stopConnection()
        else:
            # Because of the unix thing
            self.__class__.__bases__[0].channelClosed(self, channel)


class SSHSession(channel.SSHChannel):
    name = b"session"

    def channelOpen(self, foo):
        log.msg(f"session {self.id} open")
        if options["agent"]:
            d = self.conn.sendRequest(
                self, b"auth-agent-req@openssh.com", b"", wantReply=1
            )
            d.addBoth(lambda x: log.msg(x))
        if options["noshell"]:
            return
        if (options["command"] and options["tty"]) or not options["notty"]:
            _enterRawMode()
        c = session.SSHSessionClient()
        if options["escape"] and not options["notty"]:
            self.escapeMode = 1
            c.dataReceived = self.handleInput
        else:
            c.dataReceived = self.write
        c.connectionLost = lambda x: self.sendEOF()
        self.stdio = stdio.StandardIO(c)
        fd = 0
        if options["subsystem"]:
            self.conn.sendRequest(self, b"subsystem", common.NS(options["command"]))
        elif options["command"]:
            if options["tty"]:
                term = os.environ["TERM"]
                winsz = fcntl.ioctl(fd, tty.TIOCGWINSZ, "12345678")
                winSize = struct.unpack("4H", winsz)
                ptyReqData = session.packRequest_pty_req(term, winSize, "")
                self.conn.sendRequest(self, b"pty-req", ptyReqData)
                signal.signal(signal.SIGWINCH, self._windowResized)
            self.conn.sendRequest(self, b"exec", common.NS(options["command"]))
        else:
            if not options["notty"]:
                term = os.environ["TERM"]
                winsz = fcntl.ioctl(fd, tty.TIOCGWINSZ, "12345678")
                winSize = struct.unpack("4H", winsz)
                ptyReqData = session.packRequest_pty_req(term, winSize, "")
                self.conn.sendRequest(self, b"pty-req", ptyReqData)
                signal.signal(signal.SIGWINCH, self._windowResized)
            self.conn.sendRequest(self, b"shell", b"")
            # if hasattr(conn.transport, 'transport'):
            #    conn.transport.transport.setTcpNoDelay(1)

    def handleInput(self, char):
        if char in (b"\n", b"\r"):
            self.escapeMode = 1
            self.write(char)
        elif self.escapeMode == 1 and char == options["escape"]:
            self.escapeMode = 2
        elif self.escapeMode == 2:
            self.escapeMode = 1  # So we can chain escapes together
            if char == b".":  # Disconnect
                log.msg("disconnecting from escape")
                stopConnection()
                return
            elif char == b"\x1a":  # ^Z, suspend

                def _():
                    _leaveRawMode()
                    sys.stdout.flush()
                    sys.stdin.flush()
                    os.kill(os.getpid(), signal.SIGTSTP)
                    _enterRawMode()

                reactor.callLater(0, _)
                return
            elif char == b"R":  # Rekey connection
                log.msg("rekeying connection")
                self.conn.transport.sendKexInit()
                return
            elif char == b"#":  # Display connections
                self.stdio.write(b"\r\nThe following connections are open:\r\n")
                channels = self.conn.channels.keys()
                channels.sort()
                for channelId in channels:
                    self.stdio.write(
                        networkString(
                            "  #{} {}\r\n".format(
                                channelId, self.conn.channels[channelId]
                            )
                        )
                    )
                return
            self.write(b"~" + char)
        else:
            self.escapeMode = 0
            self.write(char)

    def dataReceived(self, data):
        self.stdio.write(data)

    def extReceived(self, t, data):
        if t == connection.EXTENDED_DATA_STDERR:
            log.msg(f"got {len(data)} stderr data")
            if ioType(sys.stderr) == str:
                sys.stderr.buffer.write(data)
            else:
                sys.stderr.write(data)

    def eofReceived(self):
        log.msg("got eof")
        self.stdio.loseWriteConnection()

    def closeReceived(self):
        log.msg(f"remote side closed {self}")
        self.conn.sendClose(self)

    def closed(self):
        global old
        log.msg(f"closed {self}")
        log.msg(repr(self.conn.channels))

    def request_exit_status(self, data):
        global exitStatus
        exitStatus = int(struct.unpack(">L", data)[0])
        log.msg(f"exit status: {exitStatus}")

    def sendEOF(self):
        self.conn.sendEOF(self)

    def stopWriting(self):
        self.stdio.pauseProducing()

    def startWriting(self):
        self.stdio.resumeProducing()

    def _windowResized(self, *args):
        winsz = fcntl.ioctl(0, tty.TIOCGWINSZ, "12345678")
        winSize = struct.unpack("4H", winsz)
        newSize = winSize[1], winSize[0], winSize[2], winSize[3]
        self.conn.sendRequest(self, b"window-change", struct.pack("!4L", *newSize))


class SSHListenClientForwardingChannel(forwarding.SSHListenClientForwardingChannel):
    pass


class SSHConnectForwardingChannel(forwarding.SSHConnectForwardingChannel):
    pass


def _leaveRawMode():
    global _inRawMode
    if not _inRawMode:
        return
    fd = sys.stdin.fileno()
    tty.tcsetattr(fd, tty.TCSANOW, _savedRawMode)
    _inRawMode = 0


def _enterRawMode():
    global _inRawMode, _savedRawMode
    if _inRawMode:
        return
    fd = sys.stdin.fileno()
    try:
        old = tty.tcgetattr(fd)
        new = old[:]
    except BaseException:
        log.msg("not a typewriter!")
    else:
        # iflage
        new[0] = new[0] | tty.IGNPAR
        new[0] = new[0] & ~(
            tty.ISTRIP
            | tty.INLCR
            | tty.IGNCR
            | tty.ICRNL
            | tty.IXON
            | tty.IXANY
            | tty.IXOFF
        )
        if hasattr(tty, "IUCLC"):
            new[0] = new[0] & ~tty.IUCLC

        # lflag
        new[3] = new[3] & ~(
            tty.ISIG
            | tty.ICANON
            | tty.ECHO
            | tty.ECHO
            | tty.ECHOE
            | tty.ECHOK
            | tty.ECHONL
        )
        if hasattr(tty, "IEXTEN"):
            new[3] = new[3] & ~tty.IEXTEN

        # oflag
        new[1] = new[1] & ~tty.OPOST

        new[6][tty.VMIN] = 1
        new[6][tty.VTIME] = 0

        _savedRawMode = old
        tty.tcsetattr(fd, tty.TCSANOW, new)
        # tty.setraw(fd)
        _inRawMode = 1


if __name__ == "__main__":
    run()
